decoder/vaapidecoder_h265.cpp

```diff
@@ -0,0 +1,1784 @@
+
+    m_dpb[m_dpb_count++] = fs;
+    return true;
+}
+
+/* C.5.2.2 */
+bool VaapiDecoderH265::dpb_init (VaapiDecPictureH265 * picture, H265SliceHdr *slice_hdr,  H265NalUnit * nalu)
+{
+    H265PPS *const pps = slice_hdr->pps;
+    H265SPS *const sps = pps->sps;
+
+    if (nal_is_irap (nalu->type)
+      && picture->m_NoRaslOutputFlag && !m_new_bitstream) {
+        if (nalu->type == H265_NAL_SLICE_CRA_NUT)
+            picture->m_NoOutputOfPriorPicsFlag = 1;
+        else
+            picture->m_NoOutputOfPriorPicsFlag =
+            slice_hdr->no_output_of_prior_pics_flag;
+
+        if (picture->m_NoOutputOfPriorPicsFlag)
+            dpb_clear (true);
+        else {
+            dpb_clear (false);
+            while (dpb_bump (NULL));
+        }
+    } else {
+        dpb_clear (false);
+        while ((dpb_get_num_need_output () >
+            sps->max_num_reorder_pics[sps->max_sub_layers_minus1])
+            || (sps->max_latency_increase_plus1[sps->max_sub_layers_minus1]
+            && check_latency_cnt ())
+            || (m_dpb_count >=
+            (sps->max_dec_pic_buffering_minus1[sps->max_sub_layers_minus1] +
+            1))) {
+            dpb_bump (picture);
+        }
+    }
+
+    return true;
+}
+
+Decode_Status VaapiDecoderH265::dpb_reset (uint32_t dpb_size)
+{
+    if (dpb_size > m_dpb_size) {
+        memset (&m_dpb[m_dpb_size], 0,
+        (dpb_size - m_dpb_size) * sizeof (m_dpb[0]));
+        m_dpb_size = dpb_size;
+    }
+    return true;
+}
+
+/* Detection of the first VCL NAL unit of a coded picture (7.4.2.4.5 ) */
+bool VaapiDecoderH265::is_new_picture (H265SliceHdr *slice_hdr)
+{
+    if (slice_hdr->first_slice_segment_in_pic_flag)
+        return true;
+
+    return false;
+}
+
+bool VaapiDecoderH265::has_entry_in_rps (VaapiDecPictureH265 * dpb_pic,
+    VaapiDecPictureH265 ** rps_list, uint32_t rps_list_length)
+{
+    uint32_t i;
+
+    if (!dpb_pic || !rps_list || !rps_list_length)
+        return false;
+
+    for (i = 0; i < rps_list_length; i++) {
+        if (rps_list[i] && rps_list[i]->m_poc == dpb_pic->m_poc)
+            return true;
+    }
+    return false;
+}
+
+/* the derivation process for the RPS and the picture marking */
+void VaapiDecoderH265::derive_and_mark_rps (VaapiDecPictureH265 * picture, int32_t * CurrDeltaPocMsbPresentFlag, int32_t * FollDeltaPocMsbPresentFlag)
+{
+    VaapiDecPictureH265 *dpb_pic = NULL;
+    uint32_t i;
+
+    memset (m_RefPicSetLtCurr, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetLtFoll, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStCurrBefore, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStCurrAfter, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicSetStFoll, 0, sizeof (VaapiDecPictureH265 *) * 16);
+
+    /* (8-6) */
+    for (i = 0; i < m_NumPocLtCurr; i++) {
+        if (!CurrDeltaPocMsbPresentFlag[i]) {
+            dpb_pic = dpb_get_picture (m_PocLtCurr[i], true);
+            if (dpb_pic)
```

Comment:
what is the difference between this if and a simple statement "m_RefPicSetLtCurr[i]  = dpb_pic;"